### PR TITLE
Bug 1181621 - Changing default search engine re-enables quick-search engines

### PR DESCRIPTION
**Pull request by allenngn**
To make any engine a default engine, you have to enable it, even if it was a disabled quick search engine before. When picking a new default engine, the app doesn't set the old default engine back to its original disabled state when a new default engine is chosen. Thus, previously disabled quick search engines become reenabled upon becoming a default and then becoming quick search again.

The way I tried to fix this bug was to make the app remember whether a search engine was disabled or enabled before it became a default engine. That way, a default can be set back to its original setting when it stops being a default. To do this, I added an extra key to the disabledEngines dictionary.

### CODE REVIEW

File: Client/Frontend/Browser/SearchEngines.swift [hunk 1]
```diff
@@ -51,12 +51,21 @@ class SearchEngines {
         }
 
         set(defaultEngine) {
+            let previousEngine = self.orderedEngines[0]
+            let changeKey = "BeforeCurrentWasDefault"
+            let shouldDisablePrevious = self.disabledEngineNames[changeKey] ?? false
+            if previousEngine.shortName != defaultEngine.shortName {
+                self.disabledEngineNames[changeKey] = self.disabledEngineNames[defaultEngine.shortName] ?? nil
```
Comment (thebnich) [inline Client/Frontend/Browser/SearchEngines.swift]: Creating a fake engine key to keep track of this state is a hack; I'd create a separate preference to track it.

Comment (allenngn) [inline Client/Frontend/Browser/SearchEngines.swift]: Hey just redid it. Is that better?

**Comment by allenngn** [comment]
Hey, I folded down the patch and wrote some tests. Is this okay now?

**Comment by fxios-buildasaur** [comment]
Result of integration 1
Integration took 16 seconds.
? build errors: checkout-error

**Pull request closed without merging**